### PR TITLE
Fix node that do not sync when no Public Address is set

### PR DESCRIPTION
**Pull request by NicolasDP**
will help with #860 

This address multiple items:

* put in place the initial tooling for node id authentication;
* put the required system for the non publicly reachable node to continue to receive propagation

This comes with a breaking change though...

Now the trusted peers will need to create cryptographic material to identify them selves for the other nodes (see `private_id` in the `config.yaml` for the node):

When a node wants to be publicly reachable and wants to advertise themselves as trusted peers they will have to give a public key along their public address:

to generate that key:

```
jcli key generate --type Ed25519
```

and add the output of that command in the node's config file:

```yaml
p2p:
    private_id: ed25519_sk1z30an6rtv26z8fyjdgqgcjtxt9fph7n7pw93g86ayrntx62vgz7sy4hfjs
    public_address: /ip4/x.y.z.t/tcp/8010
```

That node will them publicly announce the `public_address` and the associated public key of the `private_id`, to get the public key:

```
echo ed25519_sk1z30an6rtv26z8fyjdgqgcjtxt9fph7n7pw93g86ayrntx62vgz7sy4hfjs | jcli key to-public
```

And now to connect to that nodes user will need to either add the associated public key of the trusted peer:

```yaml
p2p:
    trusted_peers:
        - address /ip4/x.y.z.t/tcp/8010
          ed25519_pk1hs2208lr8w5lq9wmfm674d5zafaplr0xl7pml40nacg6jp2z0umsqzs3fr
```

or add it in the command line parameter:

```
jormungandr --trusted-peer /ip4/x.y.z.t/tcp/8010@ed25519_pk1hs2208lr8w5lq9wmfm674d5zafaplr0xl7pml40nacg6jp2z0umsqzs3fr --genesis-block-hash a0f92...
```

### CODE REVIEW

File: doc/configuration/introduction.md [hunk 1]
```diff
@@ -12,9 +12,12 @@ log:
...
 p2p:
   trusted_peers:
-    - "/ip4/104.24.28.11/tcp/8299"
-    - "/ip4/104.24.29.11/tcp/8299"
+    - address: "/ip4/104.24.28.11/tcp/8299"
+      id: ed25519_pk1tyz2drx5kvsp49fk0gunkl8ktlhjc976zj5erqwavnpfmrexxxnscxln75
+    - address: "/ip4/104.24.29.11/tcp/8299"
+      id: ed25519_pk13j4eata8e567xwdqp6wjeu8wa7dsut3kj0u3tgulrsmyvveq9qxqeqr3kc
   public_address: "/ip4/127.0.0.1/tcp/8080"
+  private_id: ed25519_sk1n649x7zrmt38q6dtqkvj769wru4vpkfnrppw6d83qd7jh3uhux7qwhg8q3
```
Comment (mzabaluev) [inline doc/configuration/introduction.md]: Is this the private key? Considered a low-grade secret so it's OK to put it into the config file?

File: jormungandr/src/network/p2p/topology.rs [hunk 2]
```diff
@@ -29,13 +29,15 @@ impl From<bincode::Error> for Error {
     }
 }
 
-#[derive(Clone, Debug)]
 pub struct Node(poldercast::Node);
 
+#[derive(Clone, Debug)]
+pub struct NodeData(poldercast::NodeData);
```
Comment (mzabaluev) [inline jormungandr/src/network/p2p/topology.rs]: Losing track a bit as to why there is now `Node` and `NodeData` when the useful things are now done using the latter. Some doc comments would be nice.

Comment (NicolasDP) [reply jormungandr/src/network/p2p/topology.rs]: I understand, I am differentiating the `Node`, ourselves, from the `Data` that we have on other `Node` (or ourselves).

So now, locally everyone has a `Node` and a set of `NodeData`. The `Node` contains more info than before so adding an object to superseed the `NodeData` was necessary (i.e. we don't want to gossip private keys for example).

**Comment by rinor** [comment]
:fire: simply beautiful :fire:

File: jormungandr/src/network/p2p/topology.rs (hunk 2 above)
Comment (mzabaluev) [reply jormungandr/src/network/p2p/topology.rs]: I see. Maybe a name like `SelfNode` would help understand the difference.

**Comment by hemants1** [comment]
Quick question: how does the new node want to join the network know the public ID of trusted peers?  Is the trusted peers details such as [address] and [id] published anywhere?

**Comment by mzabaluev** [comment]
@hemants1 Yes, the expectation is that the list of trusted peers for a testnet is made available _somewhere_.

**Pull request merged**